File: SymmetricBinaryTreeIter.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode *root) {
        
        if (root == NULL) {
            return true;
        }
        
        queue<TreeNode *> q;
        TreeNode *left = root->left;
        TreeNode *right = root->right;
        
        q.push(left);
        q.push(right);
        
        while (!q.empty()) {
            left = q.front();
            q.pop();
            right = q.front();
            q.pop();
            
            if (left == NULL || right == NULL) {
                if (left == NULL && right == NULL) {
                    continue;
                }
                return false;
            }
            
            if (left->val == right->val) {
                q.push(left->left);
                q.push(right->right);
                q.push(left->right);
                q.push(right->left);
                continue;
            }
            
            return false;
        }
        
        return true;
    }
    
};

File: SymmetricBinaryTreeIter.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isSymmetric(TreeNode root) {
        
        Queue<TreeNode> stack = new LinkedList<TreeNode>();
        
        if (root == null) {
            return true;
        }
        
        stack.offer(root.left);
        stack.offer(root.right);

        while (!(stack.size() == 0)) {
            
            TreeNode left = stack.poll();
            TreeNode right = stack.poll();
            
            if (left == null || right == null) {
                if (left == null && right == null) {
                    continue;
                }
                return false;
            }
            
            if (left.val != right.val) {
                return false;
            }
            
            stack.offer(left.left);
            stack.offer(right.right);
            stack.offer(left.right);
            stack.offer(right.left);
        }
        
        return true;
    }
}
